### ISSUE TITLE
mitigate missing inbound observations: stop scanning external chains if zetachain is halted

### ISSUE THREAD

**Issue by brewmaster012** [zeta-chain/node#3115]
**Describe the Bug**
During the chain halt event on Oct 30, 2024 which lasts around 6h, there were dozens of 
inbound transactions that were not observed or reported to zetacored, likely because the
network is down but zetaclient keep moving and the observation transaction accumulated
in mempool and got dropped eventually. 

This results in missed inbound observations. 

**To Reproduce**
(I have not tried this) 
Supposedly we may be able to reproduce this in the performance test where we induce
a zetachain halt and keep inbound transaction going. 

**Expected Behavior**
A better behavior is for zetaclient to receive feedback from zetacored that the zetacored
block height is not progressing (either the network is down, or the particular zetacored node
loses connectivity, database corrupt, consensus failure etc), then the zetaclient should
stop scanning external chains. 

Zetaclient persists progress of external chain observation and is expected to pick up
from where it stopped, when the zetacored issue is fixed. 

**Screenshots**

**Environment (please complete the following information):**
- OS
- Version used

**Comment by ws4charlie** [comment]
It seems we need

1. Determine if the the zetacore node is moving (height increase) or not.
2. If zetacore stops, zetaclient disable inbound?

**Comment by brewmaster012** [comment]
Yeah if the zetacored block is not moving, observation should be stopped because any reporting tx is unlikely to succeed.

**Comment by CharlieMc0** [comment]
What is the zetaclientd database used for? Can all observation be written to the DB alongside a boolean field for whether or not it was posted successfully. Zetaclient attempts to vote and once it has been voted successfully it is updated in the database. 

This way all observations are recorded even if zetacored is down and when zetacored is back we don't have to wit for zetaclient to catch up on observation. It simply posts the votes.

**Comment by kingpinXD** [comment]
In my opinion we should do both, listing in terms of priority 
1. Zetaclient can also have a new thread that checks for zetacore block times and kills the zetaclient process.
2 . If the PostVoteInbound fails, we can write the msg to the database and try to reboardcast later.

**Comment by ws4charlie** [comment]
A halt `zetacore` is one of many cases (e.g., sequence num mismatch, chain halt, unauthorized voter, not enough gas for depositAndCall executions). For this special chain halt case, having a in-memory flag `is_zetacore_stopped` in the `zetaclient` is an okay solution. A all-kill solution could be retry based and similar to the outbound:

1. Before scanning the next inbound (or next block height), `zetaclientd` need to check if it has effectively voted for the `last_inbound` (a new concept I'm thinking about) or not;
2. If the answer is YES, it continues scanning next block; if the answer is NO, it should not move forward and retry the inbound vote.

**Comment by kingpinXD** [comment]
Regarding the `last_inbound` flag, retying would also need some modifications to vote , becuase otherwise it would keep on retring and failing endlessly.
Also, the failure of the last vote does not always guarantee that the next one will fail ? There might be conditions where the vote fails becuase of execution of the deposit .

**Comment by ws4charlie** [comment]
The root cause of a missed inbound is that: `zetaclient` thinks it has successfully voted for the inbound, the but vote msg turned out failed. To solve the problem, `zetaclient` needs to check the `last_inbound` vote and ensure it is successful before scanning next, otherwise it should keep retrying.

**Comment by kingpinXD** [comment]
Based on the converstaion with @ws4charlie 

We should implement the following 

1. Add a routine to check for zetacore blocks and take stop voting if zetacore is not producing blocks. We should also keep in mind that shutting down clients would most likely trigger the, process to be started up again, and that should be handled. I am not sure if this is being handled somewhere in the code already , if its not adding this thread is trivial 

2. For the votes itself we can 
 - Either improve the current check MonitorVoteInboundResult to verify why it does not always work. Since this is called in a goroutine and not blocking, we can do additional checks if needed 
 - Implement a last_inbound check as suggested by  above, taking into account that for some failure cases, we need not wait for all inbounds to be successful ( If the contract call fails, for example )

**Comment by ws4charlie** [comment]
> We should also keep in mind that shutting down clients would most likely trigger the, process to be started up again, and that should be handled

Regarding this. The `zetaclient` needs to setup a `zetacore` health check go routine. If `zetacore` is halt, the only action `zetaclient` can take is to shut down itself. Not sure if the `zetaclientd` container will be automatically started up again or not @CharlieMc0 @CryptoFewka . Or should both `zetacore` health check and shutting down `zetaclientd` be done operationally.


> Implement a last_inbound check as suggested by above, taking into account that for some failure cases, we need not wait for all inbounds to be successful ( If the contract call fails, for example )

More specifically, if the inbound vote failure is caused by insufficient gas limit, `zetaclient` should probably skip it. Otherwise, inbound observation will get stuck in a weird situation when:

- `5/9` of the zetaclients post inbound votes successfully and continue scanning blocks.
- `4/9` of the zetaclients get stuck on endless retries (vote inbound), because the finalizing vote can't pass due to gas limit error.

**Comment by CryptoFewka** [comment]
It's ultimately up to the observer how they ensure that zetaclient and zetacore stay up and functional. We ensure ours via systemd, with a requirement that both be up, and zetacore be up first. 

There has been a deficiency in our process for a while that could result in our zetaclient being down and not automatically restarting but that has since been fixed.

**Comment by gartnera** [comment]
https://github.com/zeta-chain/node/blob/099f8df6a2cde2e1e88a4ce1e6be09a485ef92dc/zetaclient/zetacore/client_vote.go#L188-L195

is one suspicious section

**Comment by ws4charlie** [comment]
@gartnera with `last_inbound` vote checking, I guess we probably will not rely on the `MonitorVoteInboundResult` any more.

**Comment by ws4charlie** [comment]
Not sure what's the priority set for improving inbound reliability but I can always help if needed.
